ocaml/xapi/xapi_pvs_farm.ml

```diff
@@ -26,33 +26,108 @@ let introduce ~__context ~name =
+	if servers <> [] then raise Api_errors.(Server_error
 		(pvs_farm_contains_servers, List.map Ref.string_of servers));
 	Db.PVS_farm.destroy ~__context ~self
 
+(** set the name of [self] *)
 let set_name ~__context ~self ~value =
-  not_implemented "PVS_farm.set_name"
+	if proxies ~__context ~self <> [] then
+		not_allowed "can't set a name while proxies are still active"
+	else
+		Db.PVS_farm.set_name ~__context ~self ~value
+
+(** [disjoint] is [true] if two lists don't share at least one element *)
+let disjoint xs ys =
+	List.for_all (fun x -> not @@ List.mem x ys) xs
+
+(** [shared] is [true] if at least one [sr] in [srs] is shared *)
+let shared ~__context srs =
+	List.exists (fun sr -> Db.SR.get_shared ~__context ~self:sr) srs
+
+(** [sr_hosts] returns all hosts attached to a list of [sr]s. *)
+let sr_hosts ~__context srs =
+	srs
+	|> List.map (fun sr  -> Db.SR.get_PBDs ~__context ~self:sr)
+	|> List.concat
+	|> List.map (fun pbd -> Db.PBD.get_host ~__context ~self:pbd)
+
+(** [clients] returns all hosts that use the farm as a client *)
+let clients ~__context ~self =
+	proxies ~__context ~self
+	|> List.map (fun proxy -> Db.PVS_proxy.get_VIF ~__context ~self:proxy)
+	|> List.map (fun vif -> Db.VIF.get_VM ~__context ~self:vif)
+	|> List.map (fun vm  -> Db.VM.get_resident_on ~__context ~self:vm)
+
+(** add a shared SR *)
+let add_shared_cache_storage ~__context ~self ~value =
+	let cache = Db.PVS_farm.get_cache_storage ~__context ~self in
+	match cache with
+	| []    -> Db.PVS_farm.add_cache_storage ~__context ~self ~value
+	| [_]   -> not_allowed "can't add a shared SR to an existing SR"
+	| _     -> not_allowed "can't add a shared SR to a existing SRs"
+
+(* add a local SR *)
+let add_local_cache_storage ~__context ~self ~value =
+	let cache = Db.PVS_farm.get_cache_storage ~__context ~self in
+	if shared ~__context cache then
+		not_allowed "can't add SR to a cache using shared storage"
+	else if not @@ disjoint
+			(sr_hosts ~__context cache) (sr_hosts ~__context [value]) then
+		not_allowed "SR's host is already providing storage"
+	else
+		Db.PVS_farm.add_cache_storage ~__context ~self ~value
 
 let add_cache_storage ~__context ~self ~value =
-  not_implemented "PVS_farm.add_cache_storage"
+	match Db.SR.get_shared ~__context ~self:value with
+	| true  -> add_shared_cache_storage ~__context ~self ~value
+	| false -> add_local_cache_storage  ~__context ~self ~value
 
-let remove_cache_storage ~__context ~self ~value =
-  not_implemented "PVS_farm.remove_cache_storage"
+(** remove a shared SR. This should be the only one *)
+let remove_shared_cache_storage ~__context ~self ~value =
+	let cache = Db.PVS_farm.get_cache_storage ~__context ~self in
+	if not @@ List.mem value cache then
+		not_allowed "SR is not part of the farm's storage"
+	else if proxies ~__context ~self <> [] then
+		not_allowed "can't remove SR while proxies are still active"
+	else
+		assert (cache = [value]);
+		Db.PVS_farm.remove_cache_storage ~__context ~self ~value
```

Comment:
These two lines will need to be wrapped in begin/end, otherwise the final line is outside the if-statement, and will always be executed. Or just remove the assert.